include std.net
include std.core.io
include std.core.net
include std.core.exception



class NetworkException : ErrnoException

end



class Client

	int fd

	Client(int fd)
		this.fd = fd
	end

	Client(Address address)
		this.fd = connect address.rawIp.ptr, address.port
		if this.fd == -1
			throw NetworkException "Couldn't connect to " ~ address
		end
	end

	ulong read(byte[] data, ulong offset, ulong length)
		if data == null
			throw NullException "data"
		end
		if offset + length > data.length
			throw ArrayOutOfBoundsException data, offset + length
		end
		read this.fd, data.ptr + offset, length
	end

	ulong read(byte[] data)
		if data == null
			throw NullException "data"
		end
		return this.read data, 0, data.length
	end

	ulong write(byte[] data, ulong offset, ulong length)
		if data == null
			throw NullException "data"
		end
		if offset + length > data.length
			throw ArrayOutOfBoundsException data, offset + length
		end
		write this.fd, data.ptr + offset, length
	end

	ulong write(byte[] data)
		if data == null
			throw NullException "data"
		end
		return this.write data, 0, data.length
	end

end



class Server

	int fd

	Server(Address address)
		this.fd = listen address.rawIp.ptr, address.port
		if this.fd == -1
			throw NetworkException "Couldn't listen on " ~ address
		end
	end

	Client accept()
		int cfd = accept fd
		if cfd == -1
			throw NetworkException "Couldn't accept client"
		end
		return Client cfd
	end

end
